completion_aggregator/tasks/aggregation_tasks.py

```diff
@@ -210,7 +216,8 @@ def update_for_aggregator(self, block, affected_aggregators, force):
                 possible=total_possible,
                 last_modified=last_modified,
             )
-            self.aggregators[obj.block_key] = obj
+            self.updated_aggregators.append(aggregator)
+            self.aggregators[block] = aggregator
```

Comment:
I just realized that for blocks with an existing Aggregator object in `self.aggregators`, we should not be creating new objects but instead updating values of the existing objects (without saving them). Otherwise the `created` field will be reset.

Reply:
Since `pre_save` is not going to get called, `modified` also needs to be updated.

Reply:
@symbolist : Please correct me if I am wrong.

I think this is why we are using "INSERT ON DUPLICATE UPDATE" query in mysql. If there will be any object(whose value is already present in the database) and it is being inserted, it will be updated with new values for `earned`, `total`, `percent` and `last_modified`. Another point to support this, when we use `submit_completion`, which internally uses `create_or_update`. It returns a flag which tells if the new object was created or not. In the present codebase, there is no support for such case.

Also since the `Aggregator` objects are not saved, `created` and `modified` will not be populated before being saved to db.

Reply:
@taranjeet The thing is that `created` and `modified` fields have a default value set to the current time: https://github.com/jazzband/django-model-utils/blob/master/model_utils/fields.py#L12-L36.

Now, if I create a new Aggregator object for an existing `(user, course_key, block_key, aggregation_name)` object, its `created` and `modified` values are not going to be what they are in the database. So the Aggregator objects that are going to be passed to `bulk_create_or_update` and are going to be set in `self.aggregators` are going to be invalid. The difference with `create_or_update` is that the later is passed data with differentiation between the key fields and others and as you noted returns an object that can put into `self.aggregators`. Yes, this does work here because of how `bulk_create_or_update` is internally implemented but callers should not assume anything about that and because they are not currently looking at the `created` or `modified` fields.

As far as I see the only thing needed here is adding a check that if there is an existing object, four of its fields are updated. Otherwise a new object is created as is happening here already.

Reply:
@symbolist Thanks for the detailed info. :)

I have updated the codebase.